### DISCUSSION TITLE
Support of non-socket file descriptors like those of a /dev/tty...

### DISCUSSION THREAD

**Discussion by jesskomput** [cesanta/mongoose#1989] [Ask questions, get answers] [answered]
Hello, my context is mongoose running on a Linux.

Is there already a way to add a device like a /dev/ttyUSB0 to the [poll|select]ed  list of mg_connections?.
This would allows to communicate through HTTPS with for example a controller board like Arduino, ESP, ... without the need of multi-threading.

If not, I can suggest some simple source mods in a "feature request" post, my tests seem to work although there may be things missing.

**Comment by scaprile** [comment]
Could you ellaborate on your use case, please ? I can't follow you.
You can do something like what I think you want to do with the multi-threading example, why can't you use multi-threading in Linux ? You could also just check for serial data in the event handler with MG_EV_POLL.

**Reply by jesskomput** [reply]
My use case:  a web-page with some javascript sending commands/getting-events via a WebSocket connected mongoose server to/from an Arduino-UNO.

I know it's doable with multi-threading, but I was used to be able to add file descriptors got from _open()_ calls to the _select_ in various frameworks hiding the _select/poll_ from the programmer.
So I had a look at the source code and quickly found _mg_wrapfd()_ which I thought would do it.

Some debugging helped to find why this didn't work, and the addition of a flag _if_tty_ in the _struct mg_connection_ plus two conditional calls of _read()_ and _write()_ seemed to solve it.

Again, a mod. just for simplicity, no inter-thread comm, no polling, no second select... (multi-threading is for experienced C developers).

Also,  as you suggest, I will look in detail what can be done with _MG_EV_POLL_, I suspect it's kind of polling.

**Comment by scaprile** [comment, marked as answer]
So you have a Linux machine running Mongoose as a websocket server and for some reason you connect an 8-bit hardware with a serial port to one of that LInux machine serial ports.
Somehow and for some reason you need to exchange something with the 8-bitter when something happens somewhere.

On every call to `mg_mgr_poll()`, when the select()/poll() call returns, all outstanding connections are polled and receive an `MG_EV_POLL` event.
On your websocket connection:
- When it receives an `MG_EV_WS_MSG` event and you need to exchange something with the 8-bitter, you write to your serial port.
- When it receives an `MG_EV_POLL` event call, you check your serial port for data from the 8-bitter and read from it.
